### PR TITLE
feat: implement durability and resumable workflow

### LINKED ISSUE #85: Add opt-in file-based durability and crash recovery (Resume)
## Goal

Add **opt-in, zero-infrastructure crash recovery** to automa: a workflow can survive a process crash / restart / power loss and then either **resume forward** from where it stopped or **compensate backward**, without re-running completed work and without any external server or database.

Full design: [`docs/durability.md`](../blob/main/docs/durability.md).

## Why / use cases

For single-process, short-to-medium, statically-structured workflows where steps are expensive or externally observable and partial progress is worth preserving across an interruption — provisioning tools, multi-stage data migrations, installers, machine-setup orchestration. Persistence is a single local file; nothing to operate.

## Approach (summary)

- Persist **state** (a snapshot of where the saga is), not an event history. Recover by re-attaching the same workflow definition (caller's code) to the persisted state and continuing from a cursor.
- Snapshot journal (not WAL) — one file per run — composed largely from already-serializable types (`NamespacedStateBag`, `Report`, `TypeMode`).
- Durable write = marshal → temp file → `fsync` → atomic `rename`.
- Write-ahead before each `Execute`, commit after; per-compensation commit during rollback.
- New `ResumeWorkflow(ctx, wb, journalPath)` entry point; `validateTopology` refuses to resume if step IDs/order changed.

## Scope of this issue (core behaviour only)

- [ ] `Journal` / `StepJournal` types with `Version` field (`docs/durability.md` schema)
- [ ] Atomic `persist()` (temp + fsync + rename) with crash-safety tests
- [ ] Journal load + corruption handling (fail loudly, never silently restart)
- [ ] Persist points wired into the existing `Execute` loop (write-ahead + commit)
- [ ] Persist points wired into `rollbackFrom` (per-compensation commit, skip already-`Compensated`)
- [ ] `WithJournal(path)` (or equivalent) on `WorkflowBuilder`
- [ ] `ResumeWorkflow` entry point: resume-forward and compensate-backward paths
- [ ] `validateTopology` (step ID/order match) with clear error on mismatch
- [ ] Idempotency / state-in-bag / topology contract documented prominently in package docs
- [ ] Tests: crash-and-resume at each phase boundary (before exec, after exec, mid-rollback), corruption, topology mismatch

## Explicit non-goals (separate future issues)

Durable timers / long sleeps, distributed/multi-process execution, signals/queries, dynamic topology, append-only WAL + compaction. Also deferred: timeout / retry / backoff (orthogonal, easy to add once the journal exists).

## Contract this introduces for users

Steps and compensations must be idempotent (a `Started`-but-not-`Completed` step is re-run on resume); resume-relevant data must live in the state bag, not in closures; topology must be reconstructible from code + persisted state. See design doc.

## Open questions

See "Open questions" in `docs/durability.md` (storage backend, journal lifecycle/pruning, builder API shape, handling of additive topology changes).

### LINKED ISSUE #86: Durability S1: journal types & atomic persistence
Part of #85 — Story 1 of 4 (foundation; no user-visible behavior change).

Establish the on-disk schema and the atomic write/read primitives that the rest of the durability feature builds on.

### Scope
- Types: \`Journal\`, \`StepJournal\`, \`Phase\` (\`forward\`/\`compensating\`/\`done\`), \`StepState\` (\`pending\`/\`started\`/\`completed\`/\`failed\`/\`compensated\`).
- JSON marshal/unmarshal round-trip for the journal (reusing the already-serializable \`NamespacedStateBag\`, \`Report\`, \`TypeMode\`).
- \`persist(path)\`: marshal → temp file → \`fsync\` → atomic \`rename\`.
- \`loadJournal(path)\`: load + decode; corrupt file fails loudly (no silent restart).
- \`validateTopology(wf, journal)\`: compare rebuilt step IDs and order against the journal.

### Tests
- Journal JSON round-trips losslessly.
- Atomic write: a reader always sees old-or-new, never a torn file (simulate mid-write).
- Topology mismatch (IDs/order changed) is rejected with a clear error.

### Out of scope
Wiring into the execution loop (Story 2) and resume (Story 3).

Design: \`docs/durability.md\`.

### LINKED ISSUE #87: Durability S2: wire persist points into Execute/rollback
Part of #85 — Story 2 of 4 (opt-in, additive). Depends on Story 1.

Make a workflow produce a correct journal as it runs. No resume yet — this story is fully testable by inspecting the on-disk journal after each transition.

### Scope
- \`WithJournal(path)\` on the workflow builder (opt-in; default behavior unchanged, writes nothing).
- Forward loop (\`workflow.go\` \`Execute\`): write-ahead record (\`started\` + cursor) **before** the side effect; commit record (\`completed\`/\`failed\` + snapshot + report + global) **after**.
- Rollback loop (\`rollbackFrom\`): per-compensation write (\`compensated\`); set \`Phase = compensating\` on entry.
- \`Phase = done\` (or delete journal) on completion.

### Tests
- After each forward step, the on-disk journal matches expected state/cursor/snapshot.
- A failure with \`RollbackOnError\` flips phase to \`compensating\` and records each \`compensated\` step.
- Workflows without \`WithJournal\` write nothing to disk (backward compat).

### Note
Persistence depends on per-step state snapshots being correct — the \`WithState\` shallow-copy / \`prepare\`-not-called issues (#80–#84) may need to land first.

Design: \`docs/durability.md\`.

### LINKED ISSUE #88: Durability S3: ResumeWorkflow (forward + compensate)
Part of #85 — Story 3 of 4 (the payoff). Depends on Stories 1 & 2.

Add the public \`ResumeWorkflow\` entry point that rehydrates a journal onto a re-supplied workflow definition and continues.

### Scope
- \`ResumeWorkflow(ctx, wb, journalPath) *Report\`.
- Missing journal → start a fresh run. Corrupt journal → fail loudly (never silently restart; risks double side effects).
- \`validateTopology\` gate before proceeding.
- \`PhaseForward\`: re-run any \`started\`-but-not-\`completed\` step (idempotency contract), then continue from first incomplete index.
- \`PhaseCompensating\`: continue rollback from cursor, skipping \`compensated\`.
- \`PhaseDone\`: return final report.

### Tests (the crash-and-resume scenarios from the doc)
- Crash after a clean commit → completed steps skipped, remaining steps run.
- Crash with a \`started\`-but-not-\`completed\` step → that step re-runs exactly once.
- Crash mid-compensation → rollback completes, already-compensated steps skipped.
- Topology changed between crash and resume → refused.

This story holds the only subtle correctness (the \`started\` re-run) — concentrate review here.

Design: \`docs/durability.md\`.

### LINKED ISSUE #89: Durability S4: docs & resumable example
Part of #85 — Story 4 of 4 (adoption). Depends on Story 3.

### Scope
- Promote \`docs/durability.md\` status from **proposed** → **implemented**; reconcile any API drift (e.g. final \`WithJournal\` shape, run-id path namespacing).
- Add \`examples/resumable_setup/\` mirroring \`examples/setup_local\` but with a journal: run it, kill it mid-workflow, re-run to demonstrate resume.
- Document the step-author contract (idempotent steps, idempotent compensations, resume data in the state bag, reconstructible topology) prominently in README/architecture docs.

Design: \`docs/durability.md\`.

### PR DESCRIPTION
**Pull request by leninmehedy**
## Description

This pull request implements and documents durable workflow (crash recovery) support in the Automa project, and introduces a comprehensive conformance test suite for the durability feature. The most important changes include making durability an opt-in, production-ready feature, updating documentation, and adding conformance fixtures and tests to ensure correct behavior and cross-language compatibility.

**Durability (Crash Recovery) Feature Release**

* `README.md`, `docs/durability.md`: Durability (crash recovery) is now implemented and documented as a production feature, not just a proposal. Workflows can opt in to durability by specifying a journal path and using `ResumeWorkflow` to recover from crashes, restarts, or power loss. The documentation now explains the durability API, design rationale, and step author contract. [[1]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R13) [[2]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R27-R74) [[3]](diffhunk://#diff-eef95cc35414311be7abec988101dd5e7324a9b09d826d8f72026b9cba44c97eL3-R10) [[4]](diffhunk://#diff-eef95cc35414311be7abec988101dd5e7324a9b09d826d8f72026b9cba44c97eL198-R205) [[5]](diffhunk://#diff-eef95cc35414311be7abec988101dd5e7324a9b09d826d8f72026b9cba44c97eR241-R247)

* `README.md`, `docs/durability.md`: The durability section now details how to use the feature, the requirements for step authors, and the guarantees provided. The design doc is updated to match the implemented API and references the normative durability spec. [[1]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R27-R74) [[2]](diffhunk://#diff-eef95cc35414311be7abec988101dd5e7324a9b09d826d8f72026b9cba44c97eL3-R10) [[3]](diffhunk://#diff-eef95cc35414311be7abec988101dd5e7324a9b09d826d8f72026b9cba44c97eL198-R205) [[4]](diffhunk://#diff-eef95cc35414311be7abec988101dd5e7324a9b09d826d8f72026b9cba44c97eR241-R247)

**Conformance Testing for Durability**

* [`conformance_test.go`](diffhunk://#diff-be4d99d66262dc3750603ef688c69733cfb5ee8f882144967fcf6272da7b080eR26): Adds a suite of conformance tests that load durability journal fixtures (both roundtrip and resume-classification), verifying that the implementation matches the durability spec for serialization and resume behavior. The tests cover both schema round-trip and correct resume logic (which steps re-execute, which compensate, and final workflow status). [[1]](diffhunk://#diff-be4d99d66262dc3750603ef688c69733cfb5ee8f882144967fcf6272da7b080eR26) [[2]](diffhunk://#diff-be4d99d66262dc3750603ef688c69733cfb5ee8f882144967fcf6272da7b080eR314-R418)

* [`docs/spec/conformance/README.md`](diffhunk://#diff-9ffa6d01e3f43e638efc9d400be76cfd736e43000933b779035d472833a40cc6L22-R22): Documents the new `journal/` conformance fixtures, explaining their structure, purpose, and how they verify schema and resume decision agreement. [[1]](diffhunk://#diff-9ffa6d01e3f43e638efc9d400be76cfd736e43000933b779035d472833a40cc6L22-R22) [[2]](diffhunk://#diff-9ffa6d01e3f43e638efc9d400be76cfd736e43000933b779035d472833a40cc6R109-R159)

* [`docs/spec/conformance/journal/*.json`](diffhunk://#diff-beda2a584a20e6afdc575e1db354ba1c1fa7cacfb7f5c0185a47de925e8110c9R1-R30): Adds a comprehensive set of durability journal fixtures, including both roundtrip (schema) and resume (behavioral) cases, covering forward progress, compensation, nested workflows, and terminal states. [[1]](diffhunk://#diff-beda2a584a20e6afdc575e1db354ba1c1fa7cacfb7f5c0185a47de925e8110c9R1-R30) [[2]](diffhunk://#diff-1a4f40c8a461f6f322c8f0411ce2ad5218fbd397f1ce2189c564705fe1bc5457R1-R19) [[3]](diffhunk://#diff-909a3a696407a928e3d0ab2c72f80e67e967a325562a8717766d6cc4810f3824R1-R23) [[4]](diffhunk://#diff-e65e18d735cb6aa39bbb2f907e794a7f072e75acc566b72ff9d4242acb1d8bd6R1-R37) [[5]](diffhunk://#diff-1e62b5529bbd3d519292cc9198256db97cc2e013e9e88841b4d244b4b52403afR1-R24) [[6]](diffhunk://#diff-7f4be03b532832705a548872dbdb262cb488c6a6ec69429a34ac644014173ebeR1-R23)

**Resolved Design Questions**

* [`docs/durability.md`](diffhunk://#diff-eef95cc35414311be7abec988101dd5e7324a9b09d826d8f72026b9cba44c97eL323-R358): Updates the design doc to clarify and resolve previously open questions about storage backend, journal lifecycle, path namespacing, and topology validation. The only deferred extension is robust forward resume for nested workflows across schema changes.

---

These changes ensure that Automa's durability feature is robust, well-documented, and verifiable against a normative spec, with a test suite that can be used for cross-language compatibility.

### Related Issues

* Closes #85 
* Closes #86 
* Closes #87 
* Closes #88 
* Closes #89 
* Related to #

### Pull request (PR) checklist

* \[ ] This PR added tests (unit, integration, and/or end-to-end)
* \[ ] This PR updated documentation
* \[ ] This PR added no TODOs or commented out code
* \[ ] This PR has no breaking changes
* \[ ] Any technical debt has been documented as a separate issue and linked to this PR
* \[ ] Any `go.mod` and/or `go.sum` changes have been explained to and approved by a repository manager
* \[ ] All related issues have been linked to this PR
* \[ ] All changes in this PR are included in the description
* \[ ] When this PR merges the commits will be squashed and the title will be used as the commit message, the 'commit
  message guidelines' below have been followed

### Testing

* \[ ] This PR added unit tests
* \[ ] This PR added integration/end-to-end tests
* \[ ] These changes required manual testing that is documented below
* \[ ] Anything not tested is documented

The following manual testing was done:

* TBD

The following was not tested:

* TBD

<details>
<summary>
Commit message guidelines
</summary>
We use 'Conventional Commits' to ensure that our commit messages are easy to read, follow a consistent format, and for automated release note generation. Please follow the guidelines below when writing your commit messages:

1. BREAKING CHANGE: a commit that has a footer BREAKING CHANGE:, or appends a ! after the type/scope, introduces a
   breaking API change (correlating with MAJOR in Semantic Versioning). A BREAKING CHANGE can be part of commits of any
   type. NOTE: currently breaking changes will only bump the MAJOR version.
2. The title is prefixed with one of the following:

| Prefix    | Description                                         | Semantic Version Update | Captured in Release Notes |
|-----------|-----------------------------------------------------|-------------------------|---------------------------|
| feat:     | a new feature                                       | MINOR                   | Yes                       |
| fix:      | a bug fix                                           | PATCH                   | Yes                       |
| perf:     | performance                                         | PATCH                   | Yes                       |
| refactor: | code change that isn't feature or fix               | none                    | No                        |
| test:     | adding missing tests                                | none                    | No                        |
| docs:     | changes to documentation                            | none                    | Yes                       |
| build:    | changes to build process                            | none                    | No                        |
| ci:       | changes to CI configuration                         | none                    | No                        |
| style:    | formatting, missing semi-colons, etc                | none                    | No                        |
| chore:    | updating grunt tasks etc; no production code change | none                    | No                        |

</details>

### CODE REVIEW

File: journal.go [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+}
+
+// persist atomically writes the journal to path (durability-spec §3.6). A
+// reader — including a post-crash reader — observes either the complete previous
+// journal or the complete new one, never a torn file. The procedure is:
+// serialize, write to a temp file in the same directory, fsync it before the
+// rename, then atomically rename it over the target. The parent directory is
+// also fsynced so the rename itself survives power loss.
+func (j *Journal) persist(path string) error {
+	data, err := json.MarshalIndent(j, "", "  ")
```
[the remark this answers is not in the document]
Comment (leninmehedy) [reply journal.go:251]: we prefering indentation so that debugging is easier.
